### PR TITLE
Game master mode

### PR DESCRIPTION
**Pull request by zond**
@tttppp Please take a look, this is untested, but should sort-of provide game pausing functionality for the game master. Next steps are to provide replace player functionality as well, which shouldn't be much harder, except that the stats will be messed up - which might be OK since this is private games only.

### CODE REVIEW

File: game/game.go [hunk 1]
```diff
@@ -548,8 +582,39 @@ func (g *Game) Leavable() bool {
 	return !g.Started
 }
 
-func (g *Game) Joinable() bool {
-	return !g.Closed && g.NMembers < len(variants.Variants[g.Variant].Nations) && len(g.ActiveBans) == 0 && len(g.FailedRequirements) == 0
+func (g *Game) IsGameMasterInvited(email string) bool {
```
Comment (tttppp) [inline game/game.go:585]: Or `IsInvitedByGameMaster`?

File: game/game.go [hunk 2]
```diff
@@ -548,8 +582,39 @@ func (g *Game) Leavable() bool {
...
+	}
+	return false
+}
+
+func (g *Game) Joinable(user *auth.User) bool {
+	if len(g.ActiveBans) > 0 || len(g.FailedRequirements) > 0 {
+		return false
+	}
+	if g.Closed || g.NMembers >= len(variants.Variants[g.Variant].Nations) {
+		if g.GameMasterEnabled && g.HasReplaceableMember() && (!g.RequireGameMasterInvitation || g.IsGameMasterInvited(user.Email)) {
```
Comment (tttppp) [inline game/game.go:608]: Can a player be "replaceable" if there's no game master?

File: game/game.go [hunk 3]
```diff
@@ -548,8 +582,39 @@ func (g *Game) Leavable() bool {
...
+			return true
+		}
+		return false
+	} else {
+		if g.GameMasterEnabled && g.RequireGameMasterInvitation && !g.IsGameMasterInvited(user.Email) {
+			return false
+		}
```
Comment (tttppp) [inline game/game.go:615]: This if statement seems redundant - we return false regardless.

File: game/game.go [hunk 4]
```diff
@@ -800,19 +977,73 @@ func AllocateNations(preferers Preferers, nations godip.Nations) ([]godip.Nation
 	return result, nil
 }
 
+func (g *Game) ValidNation(nation godip.Nation) bool {
+	for _, nation := range variants.Variants[g.Variant].Nations {
+		if nation == nation {
```
Comment (tttppp) [inline game/game.go:982]: You need to change either the variable name or the parameter name.

File: game/game.go (hunk 1 above)
Comment (zond) [reply game/game.go:585]: Thanks!

File: game/game.go (hunk 2 above)
Comment (zond) [reply game/game.go]: I guess it's the beginning of more complex features, but it wreaks havoc (as you have mentioned) on the stats, so I'd prefer the havoc to be contained to private games, which game master games are at the moment.

WDYT?

File: game/game.go (hunk 3 above)
Comment (zond) [reply game/game.go:615]: Ah, the other option should have been true. Good catch :D

File: game/game.go (hunk 4 above)
Comment (zond) [reply game/game.go:976]: Thanks!

**Comment by zond** [comment]
Alright, I have finished the implementation, and created tests for everything except the different nation allocations when using game masters that have preallocated some nations.

**Comment by zond** [comment]
Aaand now the preallocation is tested too.

File: game/game.go (hunk 2 above)
Comment (tttppp) [reply game/game.go]: That's great, just wanted to check you'd thought about it

Comment (zond) [reply game/game.go]: 👍

**Review by tttppp — APPROVED** [review]
LGTM

**Pull request merged**